src/Symfony/Component/Security/Csrf/DoubleSubmitCsrfTokenManager.php

```diff
@@ -0,0 +1,268 @@
+            return false;
+        }
+
+        if (null === $isValidOrigin && null === $isValidDoubleSubmit) {
+            $this->logger?->warning('CSRF validation failed: double-submit and origin info not found.');
+
+            return false;
+        }
+
+        // Opportunistically lookup at the session for a previous CSRF validation strategy
+        $session = $request->hasPreviousSession() ? $request->getSession() : null;
+        $usageIndexValue = $session instanceof Session ? $usageIndexReference = &$session->getUsageIndex() : 0;
+        $usageIndexReference = \PHP_INT_MIN;
+        $previousCsrfProtection = (int) $session?->get($this->cookieName);
+        $usageIndexReference = $usageIndexValue;
+        $shift = $request->isMethodSafe() ? 8 : 0;
+
+        if ($previousCsrfProtection) {
+            if (!$isValidOrigin && (1 & ($previousCsrfProtection >> $shift))) {
+                $this->logger?->warning('CSRF validation failed: origin info was used in a previous request but is now missing.');
+
+                return false;
+            }
+
+            if (!$isValidDoubleSubmit && (2 & ($previousCsrfProtection >> $shift))) {
+                $this->logger?->warning('CSRF validation failed: double-submit info was used in a previous request but is now missing.');
+
+                return false;
+            }
+        }
+
+        if ($isValidOrigin && $isValidDoubleSubmit) {
+            $csrfProtection = 3;
+            $this->logger?->debug('CSRF validation accepted using both origin and double-submit info.');
+        } elseif ($isValidOrigin) {
+            $csrfProtection = 1;
+            $this->logger?->debug('CSRF validation accepted using origin info.');
+        } else {
+            $csrfProtection = 2;
+            $this->logger?->debug('CSRF validation accepted using double-submit info.');
+        }
+
+        if (1 & $csrfProtection) {
+            // Persist valid origin for both safe and non-safe requests
+            $previousCsrfProtection |= 1 & (1 << 8);
+        }
+
+        $request->attributes->set($this->cookieName, ($csrfProtection << $shift) | $previousCsrfProtection);
+
+        return true;
+    }
+
+    public function clearCookies(Request $request, Response $response): void
+    {
+        if (!$request->attributes->has($this->cookieName)) {
+            return;
+        }
+
+        $cookieName = ($request->isSecure() ? '__Host-' : '').$this->cookieName;
+
+        foreach ($request->cookies->all() as $name => $value) {
+            if ($this->cookieName === $value && str_starts_with($name, $cookieName.'_')) {
+                $response->headers->clearCookie($name, '/', null, $request->isSecure(), false, 'strict');
+            }
+        }
+    }
+
+    public function persistStrategy(Request $request): void
+    {
+        if ($request->hasSession(true) && $request->attributes->has($this->cookieName)) {
```

Comment:
```suggestion
        if ($request->hasSession(skipIfUninitialized: true) && $request->attributes->has($this->cookieName)) {
```

so its "self explanatory" from outside when reading ? 
also why not an early return like the above `clearCookies` ?

Reply:
argument names are not covered by the Symfony backward compatibility policy, so this is a no-go.

Reply:
**suggestion:** Another way to make it "self explanatory".

Extract method.

Result usage
```php
If ($this->requestHasInitializedSession($request)) {
```